feat(learner): full page chat modal

nicholasjjlim opened the pull request:
Closes #122 

## 🚀 Summary

<!--
Provide a clear and concise description of the changes you're making.
What problem does this solve? What is the motivation behind this change?
-->
This PR adds the full page chat modal when user clicks on the floating chat. This allows user to interact with the AI chat to find out more about the learning units that they want to find out more about.

## ✏️ Changes

<!--
List the specific changes you've made. For example:
- Added new feature X
- Fixed bug in Y
- Updated documentation for Z
-->
- Added new `ChatModal` component that contains all the features within the chat itself
- Added a new `darkSlate` variant on the `Badge` component 
- Renamed `isModalOpen` to `isPlayerModalOpen` to have better clarity on which modal is opened

## 🗒️  Notes
- This PR only targets the UI interfaces and interactions. 
- Data has been hardcoded since backend has yet to be integrated. 
- Certain behaviours are mocked to get a sense of the behaviour of the chat

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,222 @@
+        <button
+          onclick={onclose}
+          class="cursor-pointer rounded-full p-4 transition-colors hover:bg-slate-50"
+        >
```
Hover should be `bg-slate-200`.

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,222 @@
+        </textarea>
+
+        <button
+          class="cursor-pointer rounded-full bg-slate-950 p-4 text-white transition-colors hover:bg-slate-900/90 disabled:pointer-events-none disabled:bg-slate-900/50"
```
Hover a bit too light, can't really see the difference. Would suggest 70 or 80.

nicholasjjlim replied:
will go with 70 here

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,231 @@
+          onclick={onclose}
+          class="rounded-full p-4 transition-colors hover:bg-slate-200 focus-visible:outline-dashed focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-slate-950"
+        >
+          <ArrowLeft />
```
Can update to `ChevronDown` since it transitions downwards instead of back, can help to also update for Podcast Player too.

nicholasjjlim replied:
updated for both player and chat modals

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,231 @@
+
+    <!-- TODO: temporary hardcode height for now. To relook at how to set this height without hardcoding an arbitrary height -->
+    <div class="h-[calc(100vh-180px)] overflow-y-auto">
+      <div class="flex flex-col gap-y-4 px-1">
```
Is this `px-1` necessary?

nicholasjjlim replied:
is to account for the focus. else the edge will look off

<img width="1019" height="266" alt="image" src="https://github.com/user-attachments/assets/0ca434e7-5337-495a-aa99-2ea3d4398a95" />

kellylimmm replied:
Feels weird to add a `px-1` just for focus effect. 

@YimingIsCOLD thoughts?

kellylimmm commented:
Two things to note:
1. The text area does not go back to it's original size immediately after a long text has been sent, it only goes back after focusing on the element.
2. If there is a long text with no spacing involved, the text does not wrap itself to the next line. Instead, it will overflow.

nicholasjjlim commented:
> Two things to note:
> 
> 1. The text area does not go back to it's original size immediately after a long text has been sent, it only goes back after focusing on the element.
> 2. If there is a long text with no spacing involved, the text does not wrap itself to the next line. Instead, it will overflow.

1. Fixed the issue. I placed the resize function in the wrong place hence not all the behaviours are captured. Should be working properly now
2. Added `break-words` to handle this scenario.

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,234 @@
+
+  export interface Props {
+    /**
+     * A callback function that is called when the user clicks on the Back button.
```
Can update `Back` to `Close`.

kellylimmm commented on apps/learner/src/lib/components/ChatModal/ChatModal.svelte:
```diff
@@ -0,0 +1,234 @@
+  ];
+
+  // Disable scrolling when component mounts
+  onMount(() => {
```
This feels like over abstraction to achieve stylistic consistency. 

Maybe we can have a discussion again @YimingIsCOLD

nicholasjjlim replied:
Or alternatively, suggest we move the Svelte functions below the other non-Svelte functions, which will allow us to maintain our stylistic consistency. 

@santosral @YimingIsCOLD can let us know your thoughts on this too!

santosral replied:
IMO, I think we shouldn't use inline for `onMount` lifecycle hooks because the purpose of it is to group side-effect functions.

I suggest to put the scroll function to a helper
```suggestion
  // Put this in a helper file
  const createScrollControl = (): [() => void, () => void] => {
    const enableScroll = () => {
      document.body.style.overflow = '';
      document.body.style.touchAction = '';
    };

    const disableScroll = () => {
      document.body.style.overflow = 'hidden';
      document.body.style.touchAction = 'none';
    };

    return [enableScroll, disableScroll];
  };

  onMount(() => {
    const [enableScroll, disableScroll] = createScrollControl();
    
    disableScroll();

    return () => {
      enableScroll();
    };
  });
```

We don't need for a separate `onDestroy()` hook
> If a function is returned from onMount, it will be called when the component is unmounted.

References: https://svelte.dev/docs/svelte/lifecycle-hooks#onMount

kellylimmm commented on apps/learner/src/lib/components/Badge/Badge.svelte:
```diff
@@ -3,7 +3,7 @@
 
   export interface Props extends HTMLAttributes<HTMLDivElement> {
-    variant: 'blue' | 'orange' | 'amber' | 'teal' | 'rose' | 'purple' | 'slate';
+    variant: 'blue' | 'orange' | 'amber' | 'teal' | 'rose' | 'purple' | 'slate' | 'darkSlate';
```
Can use '-' instead. 
```suggestion
    variant: 'blue' | 'orange' | 'amber' | 'teal' | 'rose' | 'purple' | 'slate' | 'slate-dark';
```

You can consider using `'slate-light'` or keep it as `slate`.

nicholasjjlim commented on apps/learner/src/lib/components/ChatView/ChatView.svelte:
```diff
@@ -0,0 +1,231 @@
+    /**
+     * A callback function that is called when the user clicks on the Close button.
+     */
+    onclose?: MouseEventHandler<HTMLButtonElement>;
```
type not needed here

nicholasjjlim commented on apps/learner/src/lib/components/ChatView/ChatView.svelte:
```diff
@@ -0,0 +1,232 @@
+  // Disable scrolling when component mounts
+  onMount(() => {
+    document.body.style.overflow = 'hidden';
+    document.body.style.touchAction = 'none';
```
Reading up a bit online and trying out on iOS simulators, main motivation is to prevent touch actions from affecting the body when the view is up, especially on safari with URL bar at the bottom. But I realise after trying out a bit more this is not sufficient to solve this issue.

I will remove this line first and create a separate issue to handle responsiveness for chat view on mobile.

nicholasjjlim commented on apps/learner/src/lib/components/ChatView/ChatView.svelte:
```diff
@@ -0,0 +1,232 @@
+
+  let target = $state<HTMLElement | null>(null);
+  let userPrompt = $state('');
+  let convo = $state<ChatMessage[]>([]);
```
will rename this to `thread`

nicholasjjlim commented on apps/learner/src/lib/components/ChatWidget/ChatWidget.svelte:
```diff
@@ -1,6 +1,20 @@
 <button
   class="inset-shadow-sm inset-shadow-slate-200 p-5.5 pointer-events-auto flex cursor-pointer items-center justify-center rounded-full bg-white/30 shadow-sm backdrop-blur-sm transition-all hover:-translate-y-0.5 hover:shadow-md"
   aria-label="Open chat"
+  {onclick}
```
Let's go with wrapper function. I feel easier to stay consistent across our app this way.
